server/src/main/java/org/apache/druid/server/http/DataSourcesResource.java

```diff
@@ -410,7 +414,74 @@ public Response getServedSegmentsInInterval(
       }
       return Response.ok(segmentIds).build();
     }
-    return getServedSegmentsInInterval(dataSourceName, full != null, theInterval::contains);
+    return getServedSegmentsInInterval(dataSource, full != null, theInterval::contains);
+  }
+
+  @GET
+  @Path("/{dataSourceName}/unused/intervals")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ResourceFilters(DatasourceResourceFilter.class)
+  public Response getIntervalsWithUnusedSegmentsOrAllUnusedSegmentsPerIntervals(
```

Comment:
I see we're following the same API patterns in this file written over a decade ago, but I'm wondering if it'd be better to move away from the pattern where an API doing multiple things and returning different shapes of responses depending on the parameters specified. For example, in this case, IMO it'd be cleaner to have two different sets of API endpoints each serving only one functionality:
1. `getIntervalsWithUnusedSegments()`
2. `getUnusedSegmentsInInterval()` where the interval is optional similar to other parameters

What do you think?

Reply:
updated, now there are 3 apis